### ISSUE TITLE
Stage1 compiler bug

### ISSUE THREAD

**Issue by Franciman** [ziglang/zig#10413]
### Zig Version

0.9.0

### Steps to Reproduce

Clone my zig project: https://github.com/Franciman/telescope 
and run `zig build`

### Expected Behavior

I expected the compiler to produce a binary

### Actual Behavior

I received the following output by the compiler:

```
broken LLVM module found: Invalid operand types for ICmp instruction
  %28 = icmp eq %machine.program.Instr %27, { i64 undef, i4 -5 }, !dbg !22206

This is a bug in the Zig compiler.thread 60708 panic:
Unable to dump stack trace: debug info stripped
telescope...The following command terminated unexpectedly:
/usr/bin/zig build-exe /home/francesco/Progetti/telescope/src/main.zig --cache-dir /home/francesco/Progetti/telescope/zig-cache --global-cache-dir /home/francesco/.cache/zig --name telescope --enable-cache
error: the following build command failed with exit code 6:
/home/francesco/Progetti/telescope/zig-cache/o/c79325c310d1136dcf4f4dacbc430d23/build /usr/bin/zig /home/francesco/Progetti/telescope /home/francesco/Progetti/telescope/zig-cache /home/francesco/.cache/zig
```

**Comment by Luukdegram** [comment]
The problem seems to be with zero-sized sentinels. This should probably be disallowed.
A smaller reproducible of the problem:

```zig
const std = @import("std");

const MyUnion = union(enum){
    u8:u8,
    void,
};

pub fn main() void {
    var my_union: MyUnion = .{.u8 = 10};
    var my_slice: [:MyUnion.void]MyUnion = &.{my_union, my_union};

    std.debug.print("Slice: {any}", .{my_slice[0..:MyUnion.void]});
}
```

**Comment by Vexu** [comment]
Fixed by #12314
